### ISSUE TITLE
First input delay

### ISSUE THREAD

**Issue by merenbach** [hlissner/doom-emacs#3399]
**What did you expect to happen?**
Situation is with a stock install of Doom Emacs in the macOS Terminal, opening a plain text file or creating a new one.

When opening a new or existing plain text file, I expect the delay between keyboard input and its reflection on the screen to be:

* Consistent
* Near-instantaneous visually

**What actually happened?**

First key input, whether it's to move the cursor (`hjkl`) or type a colon to bring up commands, or remove a line, or switch to insert mode—all work, _after_ maybe .5 seconds. This is not expected.

Subsequent keys type quickly and without this delay. This _is_ expected.

**Additional details:**
- 2019 16-inch MacBook Pro 
- macOS 10.15.5, all Apple patches applied
- `GNU Emacs 26.3` installed with `brew install emacs` (have also tried `brew cask install emacs`, which installs a GUI version of emacs 26.3 from `emacsformacosx.com`, and the GUI exhibits on the splash screen the same behavior described in this ticket)
- Fresh install of Doom-Emacs following instructions, with optional dependencies (e.g., `fd`) installed
- Set path appropriately to include `bin/doom`
- Ran `doom sync`
- No customizations beyond this

`doom doctor` does not display any issues.

I've tried (with lots of `doom sync` invocations):

* Disabling all of the modules besides `ivy` in `.doom.d/init.el` (same issue occurs even without Evil)
* Switching from `ivy` to `helm` or `ido`
* Setting `(setq doom-modeline-enable-word-count nil)` in `.doom.d/config.el`
* Enabling `(setq garbage-collection-messages t)` and `(view-echo-area-messages)` in `.doom.d/config.el`
* Enabling `(doom/toggle-profiler)` in `.doom.d/config.el`

Some findings:

* There's no message that appears to correlate with the onset of the issue.
* I *do* see a garbage collection message when the cursor does move, so it's possible that it is GC.
* Setting `(setq gc-cons-threshold 100000000)` in `.doom.d/config.el` doesn't appear to make a difference, so unless this is getting overridden I'm not inclined to blame GC.
* This appears to occur after all the deferred modules load, _but_ when Evil mode is enabled, some Evil modules do load. Since this occurs even without Evil mode, I don't want to blame Evil here.
* If I "mess up" the config, say by having the doom theme set in `.doom.d/config.el` while the `doom` module is disabled in `.doom.d/init.el`, **there is no delay**. I originally thought this meant that it was theme related, but that appears to have been a red herring.
* Tried tweaking or disabling modeline--does not appear to be related to customizations there, as far as I can tell.
* The profiler output showed decent CPU and memory use, but drilling down, I couldn't find any module activity that looked out of the ordinary.

If this is GC related, how would we know for sure, given that a log message appears after the delay, but I don't know when the GC routine _starts_?

**Steps to reproduce:**
1. Do stock install of Doom Emacs with CLI emacs in macOS Terminal
2. Open a text file
3. (optional, makes no difference) Wait up to 30 seconds for any deferred modules to load
4. Try typing and expect the TUI to respond right away
5. Once it starts responding, expect proper behavior

**System information:**
<details><pre>
((emacs
  (version . "26.3")
  (features . "NOTIFY ACL GNUTLS LIBXML2 ZLIB MODULES THREADS")
  (build . "Oct 01, 2019")
  (buildopts "--disable-dependency-tracking --disable-silent-rules --enable-locallisppath=/usr/local/share/emacs/site-lisp --infodir=/usr/local/Cellar/emacs/26.3/share/info/emacs --prefix=/usr/local/Cellar/emacs/26.3 --with-gnutls --without-x --with-xml2 --without-dbus --with-modules --without-ns --without-imagemagick")
  (windowsys . batch)
  (daemonp))
 (doom
  (version . "2.0.9")
  (build . "grafted, HEAD -> refs/heads/develop, refs/remotes/origin/develop, refs/remotes/origin/HEAD 847bbf38 2020-06-13 17:12:23 -0400")
  (dir . "~/src/dotfiles/common/link/.doom.d/"))
 (system
  (type . darwin)
  (config . "x86_64-apple-darwin19.0.0")
  (shell . "/bin/zsh")
  (uname . "Darwin 19.5.0 Darwin Kernel Version 19.5.0: Tue May 26 20:41:44 PDT 2020; root:xnu-6153.121.2~2/RELEASE_X86_64 x86_64")
  (path "~/.emacs.d/bin" "~/.local/bin" "~/bin" "~/.local/bin" "~/bin" "~/.local/bin" "~/bin" "/usr/local/bin" "/usr/local/sbin" "/usr/local/bin" "/usr/bin" "/bin" "/usr/sbin" "/sbin" "/usr/local/opt/go/libexec/bin" "/usr/local/go/bin" "~/go/bin" "~/bin" "/usr/local/Cellar/emacs/26.3/libexec/emacs/26.3/x86_64-apple-darwin19.0.0"))
 (config
  (envfile)
  (elc-files . 122)
  (modules :completion company ivy :ui doom doom-dashboard doom-quit hl-todo modeline ophints (popup +defaults) vc-gutter vi-tilde-fringe workspaces :editor (evil +everywhere) file-templates fold snippets :emacs dired electric undo vc :checkers syntax :tools (eval +overlay) lookup magit :lang emacs-lisp markdown org sh :config (default +bindings +smartparens))
  (packages "n/a")
  (unpin "n/a")
  (elpa "n/a")))
</pre></details>

**Comment by yoavm448** [comment]
Does this occur every time you open a new text file, or only on the *first* time since you opened Emacs? Doom does a ton of lazy-loading which means most of the packages related to in-file movement and inputing load only once you open a file. So opening the first file would take a little time - many packages need to load - but successive openings shouldn't take that long.

**Comment by merenbach** [comment]
@yoavm448 thank you, great question. It's _every_ time I open a new or existing text file (or any file, really, but I wanted to try to rule out the syntax checker).

I completely understand that there's lazy loading going on, and I can see that happening with the messages pane. My concern is that by the time that the document contents (if an existing file) are displayed, the cursor is showing, and the messages have stopped coming in, I'd expect to be able to start typing right away. Instead, if I do so, my first keystroke isn't getting lost, but it is taking a moment to reflect on the screen. Once that happens, subsequent keystrokes work fine.

With the understanding that this could still be something on my end, an expectation I have is that when an app displays the interface to me, it's 100% ready for user interaction. If there's an additional half-second of loading that must be done, it's simply not yet ready.

If there were some command I could put in `.doom.d/config.el` to demand-load whatever additional procedures are needed, I'd much rather emacs take another half second to start up, so that by the time all the messages have stopped I know I can start typing immediately. Otherwise it doesn't strike me as ready to use yet.

One thing I'm curious about is if anyone else has seen this behavior. Is it possibly macOS specific? Just me?

**Comment by merenbach** [comment]
Additional edit: you don't need a text file. Just run "emacs" and press `j` to try to move down one item on the splash screen. On my machine there is a delay before the cursor moves the first time, but subsequent invocations of `j` move as normal.

**Comment by yoavm448** [comment]
The last thing you described with the splash screen dashboard is lazy loading exactly.

**Comment by merenbach** [comment]
@yoavm448 thanks, that's great to have that confirmation. It seems that "lazy" loading is a bit of a misnomer if I can't use the interface until loading has continued—if it's going to have to happen before the first interaction, why not just do it automatically at load regardless?

Along with that, I'm confused why I can't find specific messages about what's being loaded lazily when I view the message pane. Is there maybe some way to increase logging verbosity?

**Comment by yoavm448** [comment]
> if it's going to have to happen before the first interaction, why not just do it automatically at load regardless?

I'm not sure.
> I'm confused why I can't find specific messages about what's being loaded lazily when I view the message pane.

I don't know if there's a proper way to do this, but running emacs with `--debug` makes doom print a few messages when *some* packages load.

**Comment by merenbach** [comment]
Thanks, @yoavm448! The above messages came from `--debug-init`. I can try out `--debug` to see if that provides more info.

**Comment by yoavm448** [comment]
> if I could somehow force-load these

You could try to `require` them manually in your config.el

> What's the advantage of tying anything to the first keystroke, rather than just doing it while the editor is initially idle?

It's because these specific packages take a long time to load. So instead of waiting 3 seconds the first time you use magit, its dependencies are loaded in the background. You can try to use `:load-incrementally` in private `use-package!` declarations for these packages maybe. (Look in the Doom source code for examples.)

**Comment by merenbach** [comment]
Thanks, @yoavm448 ! I tried adding the following to `config.el`, with no changes noted (even after running `doom sync` again, just to be sure):

```
(require 'evil-easymotion)
(require 'evil-escape)
(require 'evil-snipe)
```

Yes, that makes sense that these packages wouldn't benefit from being loaded before they're needed. I'm 100% in agreement that lazy loading can provide huge benefits.

If it _is_ these packages that are causing the issue, that brings up a question: why would easymotion be required before any sort of interactivity? Right now there doesn't seem to be an advantage to the deferral that I'm seeing.

From a quick [reading of the docs](https://github.com/hlissner/doom-emacs/blob/develop/docs/faq.org#lazy-load-more-than-everything), in fact, easymotion isn't even supposed to be loaded until it's invoked, which I think is much more in line with what you have been telling me:

> The evil-easymotion package binds many keys, none of which are available until you load the package. Instead of loading it at startup, gs is bound to a command that loads the package, populates gs, then simulates the gs key press as though those new keys had always been there.

In my experience so far, easymotion appears to be loading at first keypress, not at first invocation of `gs` (which I'm experiencing as near-instantaneous, and which is _not_ generating additional log messages).

This in turn makes me wonder if a particular hook is being invoked before it's supposed to be—to summarize, on first keystroke instead of on the first invocation of individual tools like ivy or easymotion, etc.

**Comment by leoalekseyev** [comment]
It would be a good idea to disable these packages altogether and seeing if this makes a difference. Some of them you could turn off via `(package! ... :disable)` (see the docs for proper usage); it's possible that you would need to comment out some module config.  Another approach to help investigate this is to only load vanilla doom from the sandbox  (SPC h E, C-c C-d) and see if the problem goes away.

**Comment by merenbach** [comment]
@leoalekseyev @yoavm448 thank you both for the suggestion. Some findings:

1. Before I started using `--debug-init` or `--debug`, some of my earliest notes apparently didn't account for errors or warnings that were appearing, which contributed to some reproduction difficulty on my end when tinkering with settings.
2. Disabling a bunch of Evil-related packages, I found what I was looking for, inadvertently: an error on `doom-first-input-hook`.
3. Adding `(run-hooks 'doom-first-input-hook)` to the bottom of my `config.el` appears to completely resolve the issue and not make startup noticeably slower.

I am happy to consider this resolved for now, for me. Before I close, I want to ask if there is, in fact, a possible bug here, in that...

1. The delay I was seeing occurs in a _stock_ installation (at least for me).
2. Invoking the hook manually didn't change startup time for the stock installation.

Because the main purpose of emacs is usually going to be to interact with it, I'm curious what the advantage is to having a "first input hook" instead of just executing that as part of startup, since you're going to be executing it and causing a delay within a few seconds of startup either way, so why not handle it automatically?

**Comment by ymarco** [comment]
> Invoking the hook manually didn't change startup time for the stock installation.

It did raise startup time for me. Without running the hook in my `config.el` it takes about 0.9s, with it - 1.2s.

**Comment by merenbach** [comment]
@ymarco cumulatively I believe that time would have to be spent either way, correct? I'm assuming that the vast majority of cases want to reach step 3 as quickly as possible and with as little friction as possible:

1. Open emacs and wait for initializations
2. Make first keypress and wait for additional initializations
3. See output of first keypress and interact further with emacs from here

So before we can actually use your editor, we have to wait for steps 1 _and_ 2 to complete. Presumably this is _exactly_ the same amount of time as step 1 below:

1. Open emacs, wait for initializations, automatically run first keypress hooks without waiting for keypress
2. Make first keypress, see output of first keypress, and interact further with emacs from here

In short, _unless_ we're viewing a single terminal window's worth of text and using emacs _just_ as a viewer--which sounds like a fairly niche use for a common editor--there doesn't strike me as much of a disadvantage to automatically running the hooks.

I'm happy that I was able to resolve my issues my adding a single line to the config file. Should I make a documentation pull request, maybe? This seems like the sort of thing where most people would not see any reduction in cumulative startup time, and it makes the interface more responsive.
